icicle-compiler/test/Icicle/Test/Sea/Zebra.hs

```diff
@@ -0,0 +1,453 @@
+
+                 liftIO . withCStringLen (ByteString.unpack . Zebra.unEntityId . Zebra.entityId $ entity)
+                   $ \(id_bytes, id_length) -> do
+                     e1 <- test_fleet
+                             [ argPtr id_bytes
+                             , argInt32 (fromIntegral id_length)
+                             , argPtr nullPtr
+                             , argPtr fleet_ptr ]
+
+                     when (e1 /= nullPtr) $
+                       fail "failed to configure fleet"
+
+                     e2 <- test_read_entity
+                             [ argPtr state
+                             , argPtr (Zebra.unCEntity c_entity) ]
+
+                     when (e2 /= nullPtr) $ do
+                       err <- peekCString (castPtr e2)
+                       fail $ "failed to read entity: " <> err
+
+                 -- iprogram: { *mempool, input, ... }
+                 -- input: { *chord_time, *fact_count, *tombstone, *input_start, ... }
+                 programs0_ptr <- peekWordOff fleet_ptr 4
+                 tombstones_ptr <- peekWordOff programs0_ptr 3
+
+                 struct_count <- (\x -> x - 2) . length . stateInputVars <$>
+                   hoistEither (stateOfProgram 0 (wtAttribute wt) (wtAvalancheFlat wt))
+                 buf <- liftIO $ mallocBytes (struct_count * 8)
+
+                 let
+                   input_start
+                     = 4
+
+                   facts
+                     = fmap atFact (wtFacts wt)
+
+                   nfacts
+                     = length facts
+
+                   -- slice out the input fields at the index (kind of like a transpose)
+                   -- e.g slice out the second fact:
+                   -- src:
+                   --           fact_1 fact_2 ...
+                   --   field_1        A
+                   --   field_2        B
+                   --   ...
+                   -- dst:
+                   --   [ A, B, ... ]
+                   --
+                   slice dst fact_i =
+                     forM_ [0 .. struct_count - 1] $ \field_i -> do
+                       ptr_head <- peekWordOff programs0_ptr (input_start + field_i)
+                       let ptr_src = plusPtr ptr_head (fact_i  * 8)
+                           ptr_dst = plusPtr dst      (field_i * 8)
+                       copyBytes ptr_dst ptr_src 8
+
+                   peekInputs xs 0 =
+                     return xs
+
+                   peekInputs xs index = do
+                     let fact_i = nfacts - index
+                     tombstone <- liftIO $ peekElemOff tombstones_ptr fact_i
+
+                     x <- case errorOfWord tombstone of
+                        ExceptNotAnError -> do
+                          liftIO $ slice buf fact_i
+                          VRight . snd <$> peekOutput buf 0 ty
+                        e ->
+                          pure (VLeft (VError e))
+
+                     peekInputs (x:xs) (index - 1)
```

Comment:
this `peekInputs` might be worth moving in to sea eval, wherever `peekOutput` is